crates/l1tx/src/envelope/builder.rs

```diff
@@ -7,43 +7,58 @@ use bitcoin::{
     script::PushBytesBuf,
     ScriptBuf,
 };
-use strata_primitives::l1::payload::{L1Payload, L1PayloadType};
-use tracing::*;
-
-use crate::envelope::parser::{BATCH_DATA_TAG, ROLLUP_NAME_TAG, VERSION_TAG};
+use strata_primitives::{
+    l1::payload::{L1Payload, L1PayloadType},
+    params::Params,
+};
 
 // Generates a [`ScriptBuf`] that consists of `OP_IF .. OP_ENDIF` block
 pub fn build_envelope_script(
-    envelope_data: &L1Payload,
-    // TODO: get tagnames from config
-    rollup_name: &str,
+    params: &Params,
+    payloads: &[L1Payload],
     version: u8,
 ) -> anyhow::Result<ScriptBuf> {
+    let mut bytes = Vec::new();
+    for payload in payloads {
+        let script_bytes = build_payload_envelope(params, payload, version)?;
+        bytes.extend(script_bytes);
+    }
+    Ok(ScriptBuf::from_bytes(bytes))
+}
+
+fn build_payload_envelope(
+    params: &Params,
+    payload: &L1Payload,
+    version: u8,
+) -> anyhow::Result<Vec<u8>> {
+    let tag = get_payload_type_tag(payload.payload_type(), params)?;
     let mut builder = script::Builder::new()
         .push_opcode(OP_FALSE)
         .push_opcode(OP_IF)
-        .push_slice(PushBytesBuf::try_from(ROLLUP_NAME_TAG.to_vec())?)
-        .push_slice(PushBytesBuf::try_from(rollup_name.as_bytes().to_vec())?)
-        .push_slice(PushBytesBuf::try_from(VERSION_TAG.to_vec())?)
+        .push_slice(tag)
+        // Insert version
         .push_slice(PushBytesBuf::from([version]))
-        .push_slice(PushBytesBuf::try_from(BATCH_DATA_TAG.to_vec())?)
-        .push_int(envelope_data.data().len() as i64);
+        // Insert size
+        .push_slice(PushBytesBuf::from(
+            (payload.data().len() as u32).to_be_bytes(),
+        ));
 
-    trace!(batchdata_size = %envelope_data.data().len(), "Inserting batch data");
-    for chunk in envelope_data.data().chunks(520) {
-        trace!(size=%chunk.len(), "inserting chunk");
+    // Insert actual data
+    for chunk in payload.data().chunks(520) {
         builder = builder.push_slice(PushBytesBuf::try_from(chunk.to_vec())?);
     }
     builder = builder.push_opcode(OP_ENDIF);
-
-    Ok(builder.into_script())
+    Ok(builder.as_bytes().to_vec())
 }
 
-#[allow(dead_code)]
-fn get_payload_type_tag(payload_type: &L1PayloadType) -> anyhow::Result<PushBytesBuf> {
-    let ret = match *payload_type {
-        L1PayloadType::Checkpoint => PushBytesBuf::try_from("checkpoint".as_bytes().to_vec())?,
-        L1PayloadType::Da => PushBytesBuf::try_from("da".as_bytes().to_vec())?,
-    };
-    Ok(ret)
+fn get_payload_type_tag(
+    payload_type: &L1PayloadType,
+    params: &Params,
+) -> anyhow::Result<PushBytesBuf> {
+    Ok(match *payload_type {
+        L1PayloadType::Checkpoint => {
+            PushBytesBuf::try_from(params.rollup().checkpoint_tag.as_bytes().to_vec())?
+        }
+        L1PayloadType::Da => PushBytesBuf::try_from(params.rollup().da_tag.as_bytes().to_vec())?,
+    })
```

Comment:
Just make this return `&[u8]` and then convert it to the pushbytesbuf in the caller.